### PR TITLE
[HIPPEROS] Support for aarch64, i686 and powerpc64

### PR DESCRIPTION
**Pull request by odesenfans**
Add support for all the architectures supported by the HIPPEROS RTOS.

Added support for:
- aarch64-hipperos
- i686-hipperos
- powerpc64-hipperos

### CODE REVIEW

File: gcc/config/hipperos.h [hunk 1]
```diff
@@ -4,16 +4,16 @@
 #undef TARGET_HIPPEROS
 #define TARGET_HIPPEROS 1
 
-#undef STARTFILE_SPEC
-#define STARTFILE_SPEC ""
-
-#undef ENDFILE_SPEC
-#define ENDFILE_SPEC   ""
+// #undef STARTFILE_SPEC
```
Comment (VladimirSvoboda) [inline gcc/config/hipperos.h]: Comment why it is commented.

File: gcc/config/rs6000/hipperos.h [hunk 2]
```diff
@@ -0,0 +1,337 @@
+/* Definitions for HIPPEROS for 64-bit PowerPC.
+
+   Copyright (C) 1996-2017 Free Software Foundation, Inc.
```
Comment (VladimirSvoboda) [inline gcc/config/rs6000/hipperos.h:3]: Not sure about the copyright.

File: gcc/config.gcc [hunk 3]
```diff
@@ -928,7 +928,7 @@ case ${target} in
 esac
 
 case ${target} in
-aarch64*-*-elf | aarch64*-*-fuchsia* | aarch64*-*-rtems*)
+aarch64*-*-elf | aarch64*-*-fuchsia* | aarch64*-*-rtems* | aarch64*-*-hipperos*)
```
Comment (VladimirSvoboda) [inline gcc/config.gcc]: Alphabetical order.

File: gcc/config/rs6000/hipperos.h [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+   You should have received a copy of the GNU General Public License and
+   a copy of the GCC Runtime Library Exception along with this program;
+   see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#if !defined(RS6000_BI_ARCH)
+
+#undef TARGET_64BIT
+#define TARGET_64BIT 1
```
Comment (VladimirSvoboda) [inline gcc/config/rs6000/hipperos.h:29]: Does that mean that we can't have a 32 bits PowerPC?

File: gcc/config/rs6000/t-hipperos [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+# along with GCC; see the file COPYING3.  If not see
+# <http://www.gnu.org/licenses/>.
+
+MULTILIB_OPTIONS =
+MULTILIB_DIRNAMES =
+MULTILIB_MATCHES =
+MULTILIB_EXCEPTIONS =
+MULTILIB_REQUIRED =
+
+MULTILIB_OPTIONS += mcpu=e500mc/mcpu=e5500/mcpu=e6500
```
Comment (VladimirSvoboda) [inline gcc/config/rs6000/t-hipperos:27]: We don't need e500mc

File: libgcc/config.host [hunk 6]
```diff
@@ -1043,6 +1046,21 @@ powerpc-*-eabi*)
...
+    if test $libgcc_cv_powerpc_float128 = yes; then
+        tmake_file="${tmake_file} rs6000/t-float128"
+    fi
+
+    if test $libgcc_cv_powerpc_float128_hw = yes; then
+        tmake_file="${tmake_file} rs6000/t-float128-hw"
+    fi
+
+    extra_parts="$extra_parts ecrti.o ecrtn.o ncrti.o ncrtn.o"
+    md_unwind_header=rs6000/linux-unwind.h
```
Comment (VladimirSvoboda) [inline libgcc/config.host:1062]: Why Linux?

File: libgcc/config.host [hunk 7]
```diff
@@ -1043,6 +1046,21 @@ powerpc-*-eabi*)
 	tmake_file="${tmake_file} rs6000/t-ppccomm rs6000/t-savresfgpr rs6000/t-crtstuff t-crtstuff-pic t-fdpbit"
 	extra_parts="$extra_parts crtbegin.o crtend.o crtbeginS.o crtendS.o crtbeginT.o ecrti.o ecrtn.o ncrti.o ncrtn.o"
 	;;
+powerpc*-*-hipperos*)
+    tmake_file="${tmake_file} rs6000/t-ppccomm rs6000/t-savresfgpr rs6000/t-crtstuff rs6000/t-linux t-dfprules rs6000/t-ppc64-fp t-slibgcc-libgcc"
```
Comment (VladimirSvoboda) [inline libgcc/config.host]: Why Linux? Why libgcc?

File: libgcc/config.host [hunk 8]
```diff
@@ -608,6 +608,9 @@ x86_64-*-freebsd*)
 	tmake_file="${tmake_file} i386/t-freebsd i386/t-crtstuff"
 	md_unwind_header=i386/freebsd-unwind.h
 	;;
+i[34567]86-*-hipperos*)
+	tmake_file="$tmake_file i386/t-crtstuff t-crtstuff-pic t-libgcc-pic"
```
Comment (VladimirSvoboda) [inline libgcc/config.host:612]: Why libgcc?

File: gcc/config/rs6000/hipperos.h [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+%{mcpu=601:  %{!Dppc*: %{!Dmpc*: -Dppc601}  } } \
+%{mcpu=602:  %{!Dppc*: %{!Dmpc*: -Dppc602}  } } \
+%{mcpu=603:  %{!Dppc*: %{!Dmpc*: -Dppc603}  } } \
+%{mcpu=603e: %{!Dppc*: %{!Dmpc*: -Dppc603e} } } \
+%{mcpu=604:  %{!Dppc*: %{!Dmpc*: -Dmpc604}  } } \
+%{mcpu=750:  %{!Dppc*: %{!Dmpc*: -Dmpc750}  } } \
+%{mcpu=821:  %{!Dppc*: %{!Dmpc*: -Dmpc821}  } } \
+%{mcpu=860:  %{!Dppc*: %{!Dmpc*: -Dmpc860}  } } \
+%{mcpu=8540: %{!Dppc*: %{!Dmpc*: -Dppc8540}  } } \
+%{mcpu=e6500: -D__PPC_CPU_E6500__}"
```
Comment (VladimirSvoboda) [inline gcc/config/rs6000/hipperos.h]: Missing e5500

File: gcc/config/rs6000/hipperos.h [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+
+/* Copy and paste from linux64.h and freebsd64.h */
+#undef TARGET_AIX
+#define TARGET_AIX TARGET_64BIT
+
+/* Simplified copy and paste from linux64.h and freebsd64.h */
+#undef DOT_SYMBOLS
+#define DOT_SYMBOLS 0
+
+#define TARGET_PROFILE_KERNEL profile_kernel
```
Comment (VladimirSvoboda) [inline gcc/config/rs6000/hipperos.h:58]: ?

File: gcc/config.gcc [hunk 11]
```diff
@@ -2439,6 +2442,11 @@ powerpc-*-eabi*)
 	tmake_file="rs6000/t-fprules rs6000/t-ppcgas rs6000/t-ppccomm"
 	use_gcc_stdint=wrap
 	;;
+powerpc*-*-hipperos*)
+    tm_file="rs6000/biarch64.h ${tm_file} dbxelf.h elfos.h freebsd-spec.h newlib-stdint.h rs6000/sysv4.h rs6000/default64.h rs6000/hipperos.h"
```
Comment (VladimirSvoboda) [inline gcc/config.gcc:2446]: Why freebsd?

File: gcc/config.gcc [hunk 12]
```diff
@@ -2439,6 +2442,11 @@ powerpc-*-eabi*)
...
+    extra_options="${extra_options} rs6000/sysv4.opt rs6000/linux64.opt"
```
Comment (VladimirSvoboda) [inline gcc/config.gcc:2447]: Why Linux?

File: gcc/config/rs6000/t-hipperos [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+MULTILIB_DIRNAMES += me500mc me5500 me6500
+
+MULTILIB_OPTIONS += m32/m64
```
Comment (VladimirSvoboda) [inline gcc/config/rs6000/t-hipperos:30]: I don't think we need the m32 variant.
We anyway use another compiler for U-Boot for PowerPC.

**Review by VladimirSvoboda — CHANGES_REQUESTED** [review]
WHY? JUST WHY?

File: gcc/config/rs6000/hipperos.h (hunk 4 above)
Comment (odesenfans) [reply gcc/config/rs6000/hipperos.h:29]: Technically it should be possible, I didn't try.

File: gcc/config/rs6000/t-hipperos (hunk 5 above)
Comment (odesenfans) [reply gcc/config/rs6000/t-hipperos:27]: Doesn't matter.

File: libgcc/config.host (hunk 8 above)
Comment (odesenfans) [reply libgcc/config.host:612]: GCC simply does not compile if you do not configure libgcc for the target.

File: libgcc/config.host (hunk 7 above)
Comment (odesenfans) [reply libgcc/config.host]: Copy/pasted from the Linux implementation. Just needed to make it work, did not investigate further.

File: libgcc/config.host (hunk 6 above)
Comment (odesenfans) [reply libgcc/config.host:1062]: Because Linux. Honestly, did not try to look further into this.

File: gcc/config/rs6000/t-hipperos (hunk 13 above)
Comment (odesenfans) [reply gcc/config/rs6000/t-hipperos:30]: Doesn't matter, it might be useful someday.

File: gcc/config/rs6000/hipperos.h (hunk 2 above)
Comment (odesenfans) [reply gcc/config/rs6000/hipperos.h:3]: Me neither, need to check that at some point.

File: gcc/config/rs6000/hipperos.h (hunk 10 above)
Comment (odesenfans) [reply gcc/config/rs6000/hipperos.h:58]: Magic, leave it alone.

File: gcc/config/rs6000/hipperos.h (hunk 2 above)
Comment (VladimirSvoboda) [reply gcc/config/rs6000/hipperos.h:3]: Up.

File: gcc/config/rs6000/hipperos.h (hunk 10 above)
Comment (VladimirSvoboda) [reply gcc/config/rs6000/hipperos.h:58]: Up

File: gcc/config.gcc (hunk 12 above)
Comment (VladimirSvoboda) [reply gcc/config.gcc:2447]: Up

File: gcc/config/rs6000/hipperos.h (hunk 10 above)
Comment (VladimirSvoboda) [reply gcc/config/rs6000/hipperos.h:58]: Used for tracing option "-mprofile-kernel" apparently.

File: gcc/config.gcc (hunk 11 above)
Comment (VladimirSvoboda) [reply gcc/config.gcc:2446]: The freebsd-spec.h will notably define this: `__FreeBSD__`

File: libgcc/config.host (hunk 6 above)
Comment (VladimirSvoboda) [reply libgcc/config.host:1062]: Seems ok.

File: libgcc/config.host (hunk 7 above)
Comment (VladimirSvoboda) [reply libgcc/config.host]: `I would say that we should remove rs6000/t-linux

**Review by VladimirSvoboda — APPROVED** [review]

**Pull request merged**